### ISSUE TITLE
pcapFS crashes during mount

### ISSUE THREAD

**Issue by dl8bh** [fkie-cad/pcapFS#9]
Hi,

pcapFS crashes with one of my datasets, which wireshark loads just fine:

If you need more gdb output, please guide me...

```
root@pcapfs:/srv/pcapin# capinfos test.pcap 
File name:           test.pcap
File type:           Wireshark/tcpdump/... - pcap
File encapsulation:  Linux cooked-mode capture
File timestamp precision:  microseconds (6)
Packet size limit:   file hdr: 65535 bytes
Number of packets:   1.388 k
File size:           1.203 MB
Data size:           1.181 MB
Capture duration:    238,479098 seconds
First packet time:   2019-02-22 10:25:34,048427
Last packet time:    2019-02-22 10:29:32,527525
Data byte rate:      4.955 kBps
Data bit rate:       39 Mbps
Average packet size: 850,94 bytes
Average packet rate: 5.823 packets/s
SHA256:              79487d5484585f739db27073dd8c420e46f675622f981bf4dfdebb7cd13164e6
RIPEMD160:           29d1b68bebfcff5349bfcc427e6846d042442562
SHA1:                9651421962f8ba7c0f55cbba9d3142948594b6c5
Strict time order:   False
Number of interfaces in file: 1
Interface #0 info:
                     Encapsulation = Linux cooked-mode capture (25 - linux-sll)
                     Capture length = 65535
                     Time precision = microseconds (6)
                     Time ticks per second = 1000000
                     Number of stat entries = 0
                     Number of packets = 1388760
```

```
root@pcapfs:/srv/pcapin# gdb --args pcapfs test.pcap ../pcapout/
GNU gdb (Ubuntu 8.1-0ubuntu3) 8.1.0.20180409-git
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from pcapfs...fertig.
(gdb) r
Starting program: /usr/local/bin/pcapfs test.pcap ../pcapout/
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
0x000055555568fcfa in pcpp::IPv4Layer::getIPv4Header() ()

(gdb) bt
#0  0x000055555568fcfa in pcpp::IPv4Layer::getIPv4Header() ()
#1  0x000055555568d3ca in pcapfs::TcpFile::calcIpPayload(pcpp::Packet) ()
#2  0x000055555568efeb in pcapfs::TcpFile::createVirtualFilesFromPcaps(std::vector<std::shared_ptr<pcapfs::File>, std::allocator<std::shared_ptr<pcapfs::File> > > const&) ()
#3  0x00005555555a30ef in main ()

```

**Comment by 0x4d4c** [comment]
Hey Bernhard,
would you be able to share the pcap with us?

**Comment by dl8bh** [comment]
Unfortunately, thats not possible. But I am willing to test against this dataset and provide you with everything but the actual content (that I am not allowed to share)

**Comment by 0x4d4c** [comment]
We pushed a [new branch](https://github.com/fkie-cad/pcapFS/tree/issue_%239) with some debug output. Could you check whether pcapFS still crashes and if you get a packet dump when mounting your pcap?

**Comment by dl8bh** [comment]
Yes, I will be able to check monday morning!

**Comment by dl8bh** [comment]
```
root@pcapfs:/srv/pcapin# pcapfs.9 test.pcap ../pcapout/
2019-02-25 10:24:40.819745 | error | Packet length: 1516 [Bytes], Arrival time: 2019-02-12 11:15:44.048427
Ethernet II Layer, Src: 00:XX:XX:XX:XX:XX, Dst: 00:00:00:0X:00:0X
Payload Layer, Data length: 1502 [Bytes]

terminate called after throwing an instance of 'std::runtime_error'
  what():  nullptr for ipv4 packet
Abgebrochen (Speicherabzug geschrieben)

```
```
root@pcapfs:/srv/pcapin# gdb --args pcapfs.9 test.pcap ../pcapout/
GNU gdb (Ubuntu 8.1-0ubuntu3) 8.1.0.20180409-git
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from pcapfs.9...fertig.
(gdb) r
Starting program: /usr/local/bin/pcapfs.9 test.pcap ../pcapout/
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
2019-02-25 10:24:09.210373 | error | Packet length: 1516 [Bytes], Arrival time: 2019-02-12 11:15:44.048427
Ethernet II Layer, Src: 00:XX:XX:XX:XX:XX, Dst: 00:00:00:0X:00:0X
Payload Layer, Data length: 1502 [Bytes]

terminate called after throwing an instance of 'std::runtime_error'
  what():  nullptr for ipv4 packet

Program received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
51	../sysdeps/unix/sysv/linux/raise.c: Datei oder Verzeichnis nicht gefunden.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff56e2801 in __GI_abort () at abort.c:79
#2  0x00007ffff5d378b7 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007ffff5d3da06 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff5d3da41 in std::terminate() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff5d3dc74 in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x000055555568d4e2 in pcapfs::TcpFile::calcIpPayload(pcpp::Packet) ()
#7  0x000055555568f33d in pcapfs::TcpFile::createVirtualFilesFromPcaps(std::vector<std::shared_ptr<pcapfs::File>, std::allocator<std::shared_ptr<pcapfs::File> > > const&) ()
#8  0x00005555555a30cf in main ()

```

**Comment by 0x4d4c** [comment]
Thanks for trying!

So, we found the location of the crash. Apparently, PcapPlusPlus (the lib we use to parse most of the networking layers) returns a nullptr for the packet that was dumped to the console. Could you check whether there is anything special with this packet? Is it a regular IPv4 packet or something else?

**Comment by 0x4d4c** [comment]
Fixed in master and release 0.2.1.

The crash occurred when the link layer type was not Ethernet. Thanks @dl8bh for reporting and helping to debug this!
